### PR TITLE
RAB-614: Retrieve all locales information (Query)

### PR DESCRIPTION
**Pull request by grena**
**Description (for Contributor and Core Developer)**
- Add a query in Channel's API to retrieve all activated locales & find a locale by its code
- Add related read model `Locale`
- Query is cached

**Definition Of Done (for Core Developer only)**
- [x] Tests

### CODE REVIEW

File: src/Akeneo/Channel/back/tests/Integration/Query/Sql/SqlFindLocalesIntegration.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
+<?php
+
+namespace Akeneo\Test\Channel\Integration\Query\Sql;
+
+use Akeneo\Channel\API\Query\FindLocales;
+use Akeneo\Channel\API\Query\Locale;
+use Akeneo\Test\Integration\Configuration;
+use Akeneo\Test\Integration\TestCase;
+
+final class SqlFindLocalesIntegration extends TestCase
```
Comment (pierrejolly) [inline src/Akeneo/Channel/back/tests/Integration/Query/Sql/SqlFindLocalesIntegration.php:10]: Maybe you can add a test to check if cache work properly

**Review by pierrejolly — APPROVED** [review]

File: src/Akeneo/Channel/back/Infrastructure/Query/Sql/SqlFindLocales.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+use Akeneo\Channel\API\Query\FindLocales;
+use Akeneo\Channel\API\Query\Locale;
+use Akeneo\Tool\Component\StorageUtils\Cache\CachedQueryInterface;
+use Doctrine\DBAL\Connection;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class SqlFindLocales implements FindLocales, CachedQueryInterface
```
Comment (StevenVAIDIE) [inline src/Akeneo/Channel/back/Infrastructure/Query/Sql/SqlFindLocales.php:15]:
```suggestion
final class SqlFindLocales implements FindLocales, CachedQueryInterface
```

File: src/Akeneo/Channel/back/tests/Integration/Query/Sql/SqlFindLocalesIntegration.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        $activeLocale = $this->sqlFindLocales->find('en_US');
+        $this->assertInstanceOf(Locale::class, $activeLocale);
+        $this->assertEquals('en_US', $activeLocale->getCode());
+        $this->assertEquals(true, $activeLocale->isActivated());
+
+        $unactiveLocale = $this->sqlFindLocales->find('uk_UA');
+        $this->assertInstanceOf(Locale::class, $unactiveLocale);
+        $this->assertEquals('uk_UA', $unactiveLocale->getCode());
+        $this->assertEquals(false, $unactiveLocale->isActivated());
```
Comment (StevenVAIDIE) [inline src/Akeneo/Channel/back/tests/Integration/Query/Sql/SqlFindLocalesIntegration.php:33]: You can directly check if it's what you want no ?
```suggestion
        $activeLocale = $this->sqlFindLocales->find('en_US');
        $this->assertEquals(new Locale('en_US', true), $activeLocale);

        $inactiveLocale = $this->sqlFindLocales->find('uk_UA');
        $this->assertEquals(new Locale('uk_UA', false), $inactiveLocale);

        $unknownLocale = $this->sqlFindLocales->find('unknown');
        $this->assertNull($unknownLocale);
```

File: src/Akeneo/Channel/back/tests/Specification/Infrastructure/Query/Sql/SqlFindLocalesSpec.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+
+namespace Specification\Akeneo\Channel\Infrastructure\Query\Sql;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Result;
+use PhpSpec\ObjectBehavior;
+use Prophecy\Argument;
+
+class SqlFindLocalesSpec extends ObjectBehavior
```
Comment (StevenVAIDIE) [inline src/Akeneo/Channel/back/tests/Specification/Infrastructure/Query/Sql/SqlFindLocalesSpec.php:11]: For me it's the sign that we need to split the Query in two => one class for cache and one class for query. Maybe we also need to do in the the other class, like we do in Attribute public API. Maybe we can do it in another PR

**Review by StevenVAIDIE — APPROVED** [review]

File: src/Akeneo/Channel/back/tests/Specification/Infrastructure/Query/Sql/SqlFindLocalesSpec.php (hunk 4 above)
Comment (grena) [reply src/Akeneo/Channel/back/tests/Specification/Infrastructure/Query/Sql/SqlFindLocalesSpec.php:11]: Ok I see, I'll look at how it's done in Attribute public API, I think it can be done in another PR yes, I'll add it in the current sprint.
